devdoc/umock_c_lib_requirements_with_ids.md

```diff
@@ -1202,3 +1202,87 @@ X**SRS_UMOCK_C_LIB_01_195: [** If any error occurs during the destroy_call relat
 
 **SRS_UMOCK_C_LIB_01_196: [** The type used for the return of create_call and first argument of destroy_call shall be allowed to be any type registered with umock. **]**
 XX**SRS_UMOCK_C_LIB_01_204: [** Tracking of paired calls shall not be done if the actual call to the `create_call` is using the `SetFailReturn` call modifier. **]**
+
+## real function support
+
+Sometimes it is desirable to track the mockable function calls and also call the real implementation of those functions.
+While that is possible by simply registering a hook function for each of the mockable functions, it is also rather involved and cumbersome.
+
+In order to simplify the process, the macros `MOCKABLE_INTERFACE`, `IMPLEMENT_MOCKABLE_FUNCTION` and `REGISTER_GLOBAL_INTERFACE_HOOKS` can be used.
+
+Example:
+
+Given a test dependency with the below interface:
+
+```c
+MOCKABLE_INTERFACE(test_interface,
+    FUNCTION(, int, test_dependency_1_arg, int, a),
+    FUNCTION(, int, test_dependency_2_args, int, a, int, b)
+)
+```
+
+A user would write in the unit test code:
+
+```c
+#define ENABLE_MOCKS
+
+#include "test_dependency.h"
+#include "test_dependency_real_code.c"
+
+TEST_FUNCTION(reals_are_setup_at_interface_level)
+{
+    // arrange
+    int result;
+
+    REGISTER_GLOBAL_INTERFACE_HOOKS(test_interface);
+
+    STRICT_EXPECTED_CALL(test_dependency_1_arg(45));
+
+    // act
+    result = test_dependency_1_arg(45);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, "", umock_c_get_expected_calls());
+    ASSERT_ARE_EQUAL(char_ptr, "", umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int, 42, result);
+}
+```
+
+The functions defined in `test_dependency_real_code.c` have to use `IMPLEMENT_MOCKABLE_FUNCTION`.
+The inclusion of the `test_dependency_real_code.c` file while `ENABLE_MOCKS` is defined ensures that the functions implemented in that file get automatically renamed to `real_` functions.
```

Comment:
>real_ [](start = 170, length = 5)

would be nice if there was explained this convention

Reply:
Added an extra brief description of the design to route calls to real_

---
In reply to: [293554316](https://github.com/Azure/umock-c/pull/95#discussion_r293554316) [](ancestors = 293554316)